### PR TITLE
Create a new function for generating nodes for the provider worker node tests

### PR DESCRIPTION
**Pull request by yitzhak12**
Create a new function for generating nodes for the provider worker node tests.

The function generates the node objects as follows:
- If we have 3 worker nodes in the cluster, it generates all the 3 worker nodes
- If we have more than 3 worker nodes, it generates one mgr node, 2 mon nodes, and 2 osd nodes. The nodes can overlap, but at least 3 nodes should be selected(and no more than 5 nodes).

Also, add a new test 'test_rolling_provider_worker_nodes_restart' using the function above.

### CODE REVIEW

**Review by jilju — APPROVED** [review]

File: ocs_ci/ocs/node.py [hunk 1]
```diff
@@ -2604,3 +2604,48 @@ def consumers_verification_steps_after_provider_node_replacement():
...
+    """
+    wnodes = get_worker_nodes()
+    if len(wnodes) <= 3:
+        log.info(f"Generated nodes for provider node tests: {wnodes}")
+        return get_node_objs(wnodes)
+
+    mgr_node_name = pod.get_pod_node(pod.get_mgr_pods()[0]).name
+    mon_node_names = get_mon_running_nodes()
+    if mgr_node_name in mon_node_names:
+        mon_node_names.remove(mgr_node_name)
```
Comment (DanielOsypenko) [inline ocs_ci/ocs/node.py:2630]: though it is not default but it is possible scenario that we have 2 mgr pods now. In such case we will still have ceph_node_set with same mgr_node_name and one of mon node will be in mon_node_names

File: ocs_ci/ocs/node.py [hunk 2]
```diff
@@ -2604,3 +2604,48 @@ def consumers_verification_steps_after_provider_node_replacement():
...
+        num_of_osds_to_add = 2 - len(ceph_include_osd_node_set)
+    elif len(ceph_node_set) == 2:
+        num_of_osds_to_add = 1
+    else:
+        num_of_osds_to_add = 0
+
+    osd_exclude_ceph_node_set = osd_node_set - ceph_node_set
+    osd_choices = random.sample(osd_exclude_ceph_node_set, k=num_of_osds_to_add)
+    node_choices = list(ceph_node_set) + osd_choices
+    log.info(f"Generated nodes for provider node tests: {node_choices}")
```
Comment (DanielOsypenko) [inline ocs_ci/ocs/node.py:2650]: Generated -> Will be generated; we still can fail to get node objects. get_node_objs() has `assert nodes, "Failed to get the nodes OCS objects"`

File: tests/manage/z_cluster/nodes/test_nodes_restart_ms.py [hunk 3]
```diff
@@ -295,3 +295,30 @@ def test_node_maintenance_restart(self, nodes, pvc_factory, pod_factory):
...
+            logger.info(
+                "The test is applicable only for an MS provider cluster. "
+                "Switching to the provider cluster..."
+            )
+            config.switch_to_provider()
+
+        node_count = len(get_nodes())
+        ocs_nodes = generate_nodes_for_provider_worker_node_tests()
+        for node in ocs_nodes:
+            nodes.restart_nodes(nodes=[node], wait=False)
```
[the remark this answers is not in the document]
Comment (yitzhak12) [reply tests/manage/z_cluster/nodes/test_nodes_restart_ms.py:318]: Generally, you're right. But the `wait` option in the function `restart_nodes` is not always working correctly with the Managed Service.

File: ocs_ci/ocs/node.py (hunk 2 above)
Comment (yitzhak12) [reply ocs_ci/ocs/node.py:2650]: It's a bit semantic change, but it's true.

File: tests/manage/z_cluster/nodes/test_nodes_restart_ms.py [hunk 4]
```diff
@@ -295,3 +295,30 @@ def test_node_maintenance_restart(self, nodes, pvc_factory, pod_factory):
...
+            wait_for_node_count_to_reach_status(
+                node_count=node_count, node_type=constants.WORKER_MACHINE
+            )
+            ceph_health_check(tries=40)
```
[the remark this answers is not in the document]
Comment (yitzhak12) [reply tests/manage/z_cluster/nodes/test_nodes_restart_ms.py:323]: I added a noobaa check health.

File: ocs_ci/ocs/node.py (hunk 1 above)
Comment (yitzhak12) [reply ocs_ci/ocs/node.py:2632]: Fixed.

File: ocs_ci/ocs/node.py (hunk 2 above)
Comment (yitzhak12) [reply ocs_ci/ocs/node.py]: Fixed.

**Review by DanielOsypenko — DISMISSED** [review]

File: tests/manage/z_cluster/nodes/test_nodes_restart_ms.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (yitzhak12) [reply tests/manage/z_cluster/nodes/test_nodes_restart_ms.py:323]: Actually, this check does not apply to the Managed Service, so I removed it.

**Review by jilju — APPROVED** [review]

**Review by DanielOsypenko — APPROVED** [review]

**Comment by yitzhak12** [comment]
The pr validation passed: https://ocs4-jenkins-csb-odf-qe.apps.ocp-c1.prod.psi.redhat.com/job/qe-odf-multicluster/1233/console.

File: tests/manage/z_cluster/nodes/test_nodes_restart_ms.py [hunk 5]
```diff
@@ -295,3 +295,30 @@ def test_node_maintenance_restart(self, nodes, pvc_factory, pod_factory):
 
         # Create PVCs and pods
         self.sanity_helpers.create_resources_on_ms_consumers()
+
+    @tier4b
+    @pytest.mark.polarion_id("OCS-2015")
+    def test_rolling_provider_worker_nodes_restart(self, nodes):
```
Comment (dahorak) [inline tests/manage/z_cluster/nodes/test_nodes_restart_ms.py:301]: Is this test only for Managed Service cluster or can it be used also on normal cluster?
If it is only for MS cluster, we should probably skip it on normal cluster via `@managed_service_required`.

**Review by dahorak — APPROVED** [review]

File: tests/manage/z_cluster/nodes/test_nodes_restart_ms.py (hunk 5 above)
Comment (dahorak) [reply tests/manage/z_cluster/nodes/test_nodes_restart_ms.py:301]: Missed the `@managed_service_required` mark on the class level.

Comment (yitzhak12) [reply tests/manage/z_cluster/nodes/test_nodes_restart_ms.py:301]: The decorator is on all the methods in the class: https://github.com/red-hat-storage/ocs-ci/blob/master/tests/manage/z_cluster/nodes/test_nodes_restart_ms.py#L45.

**Pull request merged**